wgengine/magicsock/magicsock_linux.go

```diff
@@ -0,0 +1,316 @@
+	// BPF and more in-depth parsing of all IPv6 packets in userspace, but
+	// on systems with a high volume of UDP that would be unacceptably slow
+	// and thus we'd rather be conservative here and possibly not receive
+	// disco packets rather than slow down the system.
+	magicsockFilterV6 = []bpf.Instruction{
+		// Check Ethernet header for EtherType = 0x86DD (IPv6)
+		bpf.LoadAbsolute{Off: 12, Size: 2},
+		bpf.JumpIf{Cond: bpf.JumpEqual, Val: 0x86DD, SkipTrue: 1, SkipFalse: 0},
+		bpf.RetConstant{Val: 0x0},
+
+		// Check "Next Header" field == UDP
+		bpf.LoadAbsolute{Off: ethHeaderSize + 6, Size: 1},
+		bpf.JumpIf{Cond: bpf.JumpEqual, Val: uint32(ipproto.UDP), SkipTrue: 0, SkipFalse: 1},
+
+		// Next header indicates we have a UDP packet, so the header
+		// offset is immediately following the IPv6 header.
+		bpf.LoadConstant{Dst: bpf.RegX, Val: ethHeaderSize + ipv6.HeaderLen},
+		bpf.Jump{Skip: 6}, // move past fragment checks
+
+		// Check if the "Next Header" field is the "Fragment" header
+		bpf.JumpIf{Cond: bpf.JumpEqual, Val: 0x2C, SkipTrue: 1, SkipFalse: 0},
+		bpf.RetConstant{Val: 0x0}, // skip packet
+
+		// If it's the fragment header, check whether the protocol is UDP
+		bpf.LoadAbsolute{Off: ethHeaderSize + ipv6.HeaderLen, Size: 1},
+		bpf.JumpIf{Cond: bpf.JumpEqual, Val: uint32(ipproto.UDP), SkipTrue: 1, SkipFalse: 0},
+		bpf.RetConstant{Val: 0x0}, // skip packet
+
+		// Fragment header's "Next Header" indicates that we have a UDP
+		// packet immediately following; the offset is immediately
+		// following both headers.
+		bpf.LoadConstant{Dst: bpf.RegX, Val: ethHeaderSize + ipv6.HeaderLen + ipv6FragmentHeaderSize},
+
+		// If we get here, we have a UDP packet; compare with our magic
+		// number. Start by loading and comparing the first 4 bytes of
+		// the UDP packet.
+		bpf.LoadIndirect{Off: udpHeaderSize, Size: 4},
+		bpf.JumpIf{Cond: bpf.JumpEqual, Val: magic1, SkipTrue: 0, SkipFalse: 3},
+
+		// Compare the next 2 bytes
+		bpf.LoadIndirect{Off: udpHeaderSize + 4, Size: 2},
+		bpf.JumpIf{Cond: bpf.JumpEqual, Val: uint32(magic2), SkipTrue: 0, SkipFalse: 1},
+
+		// Accept the whole packet
+		bpf.RetConstant{Val: 0xFFFFFFFF},
+
+		// Skip the packet
+		bpf.RetConstant{Val: 0x0},
+	}
+)
+
+// listenDisco starts listening for disco packets using an AF_PACKET socket + a
+// BPF filter. This allows us to receive disco packets even without opening the
+// firewall; see issue 3824 for more detail.
+func (c *Conn) listenDisco() {
+	if debugDisableAF_PACKET {
+		return
+	}
+	c.listenDiscoProto(&c.bpfConn4, "IPv4", magicsockFilterV4)
+	c.listenDiscoProto(&c.bpfConn6, "IPv6", magicsockFilterV6)
+}
+
+func (c *Conn) listenDiscoProto(fdptr *int, name string, filter []bpf.Instruction) {
+	// We create our socket manually here, rather than using
+	// {ipv4,ipv6}.NewPacketConn, because they don't support SOCK_RAW.
+	fd, err := unix.Socket(unix.AF_PACKET, unix.SOCK_RAW, int(htons(unix.ETH_P_ALL)))
```

Comment:
c, err := net.ListenPacket("ip4:17", "0.0.0.0")

Will make a raw UDP-only listener. As a bonus, the packets you receive won't have an ethernet envelope, meaning they'll work on non-ethernet interfaces and save you a little BPF code. And you can then use x/net/ipv4 to set the BPF filter, and end up with packet listener that integrate gracefully with the runtime netpoller.

Same for IPv6.